xplique/attributions/image_object_detector_explainer.py

```diff
@@ -0,0 +1,280 @@
+import abc
+
+import tensorflow as tf
+import numpy as np
+
+from xplique.attributions.base import BlackBoxExplainer
+from xplique.commons.tf_operations import inference_batching
+
+
+class IouCalculator:
+    """
+    Used to compute the Intersection Over Union.
+    """
+    @abc.abstractmethod
+    def intersect(self, obj1: tf.Tensor, obj2: tf.Tensor) -> tf.Tensor:
+        """
+        Compute the intersection between two batched objects
+        An object could be box, segmentation mask...
+
+        Parameters
+        ----------
+        obj1
+            object number 1. Can be a box, a segmentation masks, etc..
+        obj2
+            object number 2. Can be a box, a segmentation masks, etc..
+
+        Returns
+        -------
+        score
+            a score between [0,1] corresponding to the intersection of the 2 objects.
+        """
+        raise NotImplementedError()
+
+
+class SegmentationIouCalculator(IouCalculator):
+    """
+    Used to compute the Mask IOU.
+    """
+    def intersect(self, mask1: tf.Tensor, mask2: tf.Tensor) -> tf.Tensor:
+        """
+        Compute the intersection between two batched segmentation mask.
+        The segmentation is a boolean mask on the whole image
+
+        Parameters
+        ----------
+        mask1
+            mask of segmentation 1
+        mask2
+            mask of segmentation 2
+
+        Returns
+        -------
+        iou
+            The IOU score between mask1 and mask2
+        """
+        # pylint: disable=W0237
+        axis = np.arange(1, len(tf.shape(mask1)))
+        inter_area = tf.reduce_sum(tf.cast(tf.logical_and(mask1, mask2),
+                                    dtype=tf.float32), axis=axis)
+        union_area = tf.reduce_sum(tf.cast(tf.logical_or(mask1, mask2),
+                                    dtype=tf.float32), axis=axis)
+
+        return inter_area / tf.maximum(union_area, 1.0)
+
+
+
+class BoxIouCalculator(IouCalculator):
+    """
+    Used to compute the Bounding Box IOU.
+    """
+    EPSILON = tf.constant(1e-4)
+
+    def intersect(self, box1: tf.Tensor, box2: tf.Tensor) -> tf.Tensor:
+        """
+        Compute the intersection between two batched bounding boxes.
+        The bounding box is defined by (x1, y1, x2, y2)
+
+        Parameters
+        ----------
+        boxA
+            bounding box 1
+        boxB
+            bounding box 2
+
+        Returns
+        -------
+        iou
+            The IOU score between box1 and box2
+        """
+        # pylint: disable=W0237
+        # pylint: disable=C0103
+        # determine the (x, y)-coordinates of the intersection rectangle
+        x1 = tf.maximum(box1[..., 0], box2[..., 0])
+        y1 = tf.maximum(box1[..., 1], box2[..., 1])
+        x2 = tf.minimum(box1[..., 2], box2[..., 2])
+        y2 = tf.minimum(box1[..., 3], box2[..., 3])
+        # compute the area of intersection rectangle
+        inter_area = tf.maximum(0, x2 - x1) * tf.maximum(0, y2 - y1)

```

Comment:
Why do we need to use `tf.maximum(.)` here ? I don't follow, either we should assert at the top that for each box we have x2 > x1, or we assume it's always true, but I don't like to put max here, It could be a silent error (e.g a user alway get 0 but don't understand why).